partiql-spi/src/main/kotlin/org/partiql/spi/connector/sql/builtins/FnExtract.kt

```diff
@@ -0,0 +1,368 @@
+    }
+}
+
+//
+// Extract Minute
+//
+@OptIn(PartiQLValueExperimental::class, FnExperimental::class)
+internal object Fn_EXTRACT_MINUTE__TIME__INT32 : Fn {
+
+    override val signature = FnSignature(
+        name = "extract_minute",
+        returns = INT32,
+        parameters = listOf(
+            FnParameter("datetime", TIME),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimeValue>().value!!
+        return int32Value(v.minute)
+    }
+}
+
+@OptIn(PartiQLValueExperimental::class, FnExperimental::class)
+internal object Fn_EXTRACT_MINUTE__TIMESTAMP__INT32 : Fn {
+
+    override val signature = FnSignature(
+        name = "extract_minute",
+        returns = INT32,
+        parameters = listOf(
+            FnParameter("datetime", TIMESTAMP),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimestampValue>().value!!
+        return int32Value(v.minute)
+    }
+}
+
+//
+// Extract Second
+//
+@OptIn(PartiQLValueExperimental::class, FnExperimental::class)
+internal object Fn_EXTRACT_SECOND__TIME__DECIMAL_ARBITRARY : Fn {
+
+    override val signature = FnSignature(
+        name = "extract_second",
+        returns = DECIMAL_ARBITRARY,
+        parameters = listOf(
+            FnParameter("datetime", TIME),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimeValue>().value!!
+        return decimalValue(v.decimalSecond)
+    }
+}
+
+@OptIn(PartiQLValueExperimental::class, FnExperimental::class)
+internal object Fn_EXTRACT_SECOND__TIMESTAMP__DECIMAL_ARBITRARY : Fn {
+
+    override val signature = FnSignature(
+        name = "extract_second",
+        returns = DECIMAL_ARBITRARY,
+        parameters = listOf(
+            FnParameter("datetime", TIMESTAMP),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimestampValue>().value!!
+        return decimalValue(v.decimalSecond)
+    }
+}
+
+//
+// Extract Timezone Hour
+//
+@OptIn(PartiQLValueExperimental::class, FnExperimental::class)
+internal object Fn_EXTRACT_TIMEZONE_HOUR__TIME__INT32 : Fn {
+
+    override val signature = FnSignature(
+        name = "extract_timezone_hour",
+        returns = INT32,
+        parameters = listOf(
+            FnParameter("datetime", TIME),
+        ),
+        isNullCall = true,
+        isNullable = false,
+    )
+
+    override fun invoke(args: Array<PartiQLValue>): PartiQLValue {
+        val v = args[0].check<TimeValue>().value!!
+        return when (val tz = v.timeZone) {
+            TimeZone.UnknownTimeZone -> int32Value(0)
```

Comment:
QQ: You're the one most familiar with date time semantics, but should this be 0? While it's not a boolean value, in boolean logic, an unknown value gets coerced to a NULL value.

The Ion spec says:
> 2007-02-23T20:14:33.079Z         // The same instant in UTC ("zero" or "zulu")
2007-02-23T20:14:33.079+00:00    // The same instant, with explicit local offset
2007-02-23T20:14:33.079-00:00    // The same instant, with unknown local offset

Which shows that the timestamps themselves are equivalent. But, if you were to want to derive information from your data, I feel as if NULL would be returned upon extracting an unknown offset.

Reply:
This is a good question. 

While I don't know what should the decision be, for the time being I decided to follow the behavior from the existing pipelines, which returns zero (in form of decimal for some reasons)...

```
EXTRACT(TIMEZONE_HOUR FROM `2024-01-01T23:12:59-00:00`)
   | 
===' 
0.
```

I think for the time being we can stay with that behavior, but I will add a todo note for this line.

Great call out. Thanks.